### ISSUE TITLE
Using source directly from vql seems to result in a deadlock

### ISSUE THREAD

**Issue by ruzzle** [Velocidex/velociraptor#3746]
Compiling the latest source, I attempted to source artifact query sources directly from vql, like:

```
velociraptor query -v 'SELECT * FROM source(artifact="Windows.Forensics.SRUM/Network Connections")'
```

This however seems to result in a deadlock, without any error report. Even non-existing artifacts do not result in an error

```
velociraptor query -v 'SELECT * FROM source(artifact="foobar")'
```

Might this be a bug? Also, is this the right way to directly use (named) queries from external artifacts in direct vql?

**Comment by predictiple** [comment]
In your output you should see "Command not available when running without a frontend service. To perform administrative tasks on the command line, connect to the server using the API https://docs.velociraptor.app/docs/server_automation/server_api/"

In other words, the `query` command is for running queries as a client and `source` is a server-side plugin. To run queries on the server you have to use the API:
https://docs.velociraptor.app/docs/server_automation/server_api/#using-the-shell-for-automation

**Comment by ruzzle** [comment]
I noted this, but I also ran the same thing on velociraptor 0.7.0, which did not result in a deadlock, and exited properly via the following output:

```
velociraptor query -v 'SELECT * FROM source(artifact="foobar")'
[INFO] 2024-09-10T10:10:18Z  _    __     __           _                  __
[INFO] 2024-09-10T10:10:18Z | |  / /__  / /___  _____(_)________ _____  / /_____  _____
[INFO] 2024-09-10T10:10:18Z | | / / _ \/ / __ \/ ___/ / ___/ __ `/ __ \/ __/ __ \/ ___/
[INFO] 2024-09-10T10:10:18Z | |/ /  __/ / /_/ / /__/ / /  / /_/ / /_/ / /_/ /_/ / /
[INFO] 2024-09-10T10:10:18Z |___/\___/_/\____/\___/_/_/   \__,_/ .___/\__/\____/_/
[INFO] 2024-09-10T10:10:18Z                                   /_/
[INFO] 2024-09-10T10:10:18Z Digging deeper!                  https://www.velocidex.com
[INFO] 2024-09-10T10:10:18Z This is Velociraptor 0.7.0 built on 2023-08-28T22:37:17Z (ac56954c)
[INFO] 2024-09-10T10:10:18Z Env var VELOCIRAPTOR_API_CONFIG is not set
[INFO] 2024-09-10T10:10:18Z No embedded config - you can pack one with the `config repack` command
[INFO] 2024-09-10T10:10:18Z Env var VELOCIRAPTOR_CONFIG is not set
[INFO] 2024-09-10T10:10:18Z Setting empty config
[INFO] 2024-09-10T10:10:18Z Starting Org Manager service.
[INFO] 2024-09-10T10:10:18Z Starting services for Root Org
[INFO] 2024-09-10T10:10:18Z Starting user manager service for org
[INFO] 2024-09-10T10:10:18Z Starting Journal service for Root Org.
[INFO] 2024-09-10T10:10:18Z Starting the notification service for Root Org.
[INFO] 2024-09-10T10:10:18Z Installing Dummy inventory_service. Will download tools to temp directory.
[INFO] 2024-09-10T10:10:18Z Starting repository manager for Root Org
[INFO] 2024-09-10T10:10:18Z Loaded 372 built in artifacts in 101.789084ms
[INFO] 2024-09-10T10:10:18Z source: Artifact foobar not known
[][INFO] 2024-09-10T10:10:18Z Completed query in 204.083µs
[DEBUG] 2024-09-10T10:10:18Z Query Stats: {"RowsScanned":0,"PluginsCalled":1,"FunctionsCalled":0,"ProtocolSearch":0,"ScopeCopy":1}
[INFO] 2024-09-10T10:10:18Z Exiting notification service for Root Org!
```

**Comment by predictiple** [comment]
The behaviour was changed in https://github.com/Velocidex/velociraptor/pull/3575 to be more strict about server-only plugins. I agree that it should exit, which it does not do after this change.

There's nothing wrong with using the `source` plugin, but to do so you need to be directing your query to the API with the `--api_config` flag. As explained in the article I linked above. Otherwise your query is running in a local context which knows nothing about the server or it's sources.

**Comment by ruzzle** [comment]
Alright, thanks for explaining context! Will have a look into it 👍

**Comment by scudette** [comment]
You are right - it should not be hanging - thanks for the report!

**Comment by scudette** [comment]
Deadlock should be fixed in #3749
